File: TwitterIrcGatewayCore/AddIns/DLRIntegration/Samples/scraping.rb
# $Id$
include Misuzilla::Applications::TwitterIrcGateway

class Scraping
	@@interval = 30
	@@user_cache = {}
	@@thread = nil
	
	def self.interval=(val); @@interval = val; end
	def self.interval; @@interval; end
	def self.user_cache; @@user_cache; end
	def self.thread=(val); @@thread = val; end
	def self.thread; @@thread; end
	
	def self.prepare_user_cache
#		begin
			Session.TwitterService.GetFriends(20).each do |u|
				@@user_cache[u.ScreenName.to_s] = u
			end
#		rescue
#		end
	end
	
	def self.fetch_home
		home = CurrentSession.TwitterService.GETWithCookie("/home").to_s
		if statuses = home.scan(%r{(<li class="hentry status.*?</li>)})
			statuses.reverse.each do |status1|
				status = status1[0]
				s = Status.new

				# User
				m = status.match(%r{class="screen-name" title="([^"]+)">(.*?)</a>})
				if @@user_cache[m[2]]
					s.User = @@user_cache[m[2]]
				else
					#begin
					#	s.User = CurrentSession.TwitterService.GetUser(m[2])
					#	@@user_cache[m[2]] = s.User
					#rescue
						s.User            = User.new
						s.User.Id         = 0
						s.User.Name       = m[1]
						s.User.ScreenName = m[2]
					#end
				end
				
				# Status
				s.Source    = status.match(%r{<span>from (.*?)</span>})[1].to_s
				s.Text      = Utility::UnescapeCharReference(status.match(%r{class="entry-content">(.*?)</span>})[1].to_s.gsub(%r{<a href="(http://[^"]*)"[^>]*>.*?</a>}, '\1').gsub(/<[^>]*>/, ''))
				s.Id        = status.match(%r{id="status_(\d+)"})[1].to_i
				s.CreatedAt = Time.now
				
				# 送信
				#System::Diagnostics::Trace::WriteLine(s)
				Session.TwitterService.ProcessStatus(s, System::Action[Status].new{|s1| CurrentSession.ProcessTimelineStatus(s, false, false) })
			end
		end
	end
end

Scraping.thread = Thread.new do
	# ちゃんとThreadを終わらせないと大変残念なことになる
	dlr_addin = CurrentSession.AddInManager.GetAddIn(Misuzilla::Applications::TwitterIrcGateway::AddIns::DLRIntegration::DLRIntegrationAddIn.to_clr_type)
	dlr_addin.BeforeUnload do |sender, e|
		Scraping.interval = 0
		Scraping.thread.join
	end
	
	# ユーザ情報をキャッシュする
	Session.pre_send_message_timeline_status do |sender, e|
		if e.status.user.Id != 0
			Scraping.user_cache[e.status.user.ScreenName.to_s] = e.status.user
		end
	end
    
	# ログインする
	Session.TwitterService.CookieLogin
    
	while Scraping.interval > 0 do
		begin
			Scraping.fetch_home
		rescue => e
			# なんもしないけど
			System::Diagnostics::Trace::WriteLine(e.to_s)
		end
		sleep(Scraping.interval)
	end
end

Scraping.thread.run

File: TwitterIrcGatewayCore/AddIns/DLRIntegration/Samples/scraping.py
import clr
import re
import thread
import time

import Misuzilla.Applications.TwitterIrcGateway
import Misuzilla.Applications.TwitterIrcGateway.AddIns
import Misuzilla.Applications.TwitterIrcGateway.AddIns.Console

from System import *
from System.Threading import Thread, ThreadStart
from System.Diagnostics import Trace
from Misuzilla.Applications.TwitterIrcGateway import Status, Statuses, User, Users, Utility
from Misuzilla.Applications.TwitterIrcGateway.AddIns import IConfiguration
from Misuzilla.Applications.TwitterIrcGateway.AddIns.Console import ConsoleAddIn, Console, Context
from Misuzilla.Applications.TwitterIrcGateway.AddIns.DLRIntegration import DLRIntegrationAddIn, DLRBasicConfiguration, DLRContextHelper

class Scraping(Object):
	def __init__(self):
		Session.AddInManager.GetAddIn[DLRIntegrationAddIn]().BeforeUnload += self.onBeforeUnload
		self.thread = None
		self.interval = 30
		self.re_source = re.compile(r"<span>from (.*?)</span>")
		self.re_statuses = re.compile(r"<li class=\"hentry status.*?</li>")
		self.re_content = re.compile(r"class=\"entry-content\">(.*?)</span>")
		self.re_user = re.compile(r"class=\"screen-name\" title=\"([^\"]+)\">(.*?)</a>")
		self.re_anchor = re.compile(r"<a href=\"(http://[^\"]*)\"[^>]*>.*?</a>")
		self.re_tag = re.compile(r"<[^>]*>")
		self.re_status_id = re.compile(r"id=\"status_(\d+)\"")
		
	def start(self):
		CurrentSession.TwitterService.CookieLogin()
		self.thread = Thread(ThreadStart(self.runProc))
		self.thread.Start()

	def runProc(self):
		while self.interval > 0:
			try:
				self.fetchHome()
			except Exception, ex:
				Trace.WriteLine(ex.ToString())
			Thread.Sleep(self.interval * 1000)

	def fetchHome(self):
		home = CurrentSession.TwitterService.GETWithCookie("/home")
		statuses = self.re_statuses.findall(home)
		statuses.reverse()
		for status in statuses:
			s = Status()
			# User
			match = self.re_user.search(status)
			s.User            = User()
			s.User.Id         = 0
			s.User.Name       = match.group(1)
			s.User.ScreenName = match.group(2)
			
			# Status
			s.Source    = self.re_source.search(status).group(1)
			s.Text      = Utility.UnescapeCharReference(self.re_tag.sub(r"", self.re_anchor.sub(r"\1", self.re_content.search(status).group(1))))
			s.Id        = int(self.re_status_id.search(status).group(1), 10)
			s.CreatedAt = DateTime.Now
			
			#Trace.WriteLine(s.ToString())
			CurrentSession.TwitterService.ProcessStatus(s, Action[Status](lambda s1: CurrentSession.ProcessTimelineStatus(s1, False, False)))

	def onBeforeUnload(self, sender, e):
		self.interval = 0
		self.thread.Join()

Scraping().start()

